syntax = "proto3";

option go_package = ".;pb";

service KV {
    rpc Get (GetRequest) returns (GetReply) {}
    rpc MGet (MGetRequest) returns (MGetReply) {}
    rpc Search (SearchRequest) returns (SearchReply) {}
}

message GetRequest {
    string key = 1;
    string space = 2;
}

message GetReply {
    string value = 1;
}

message MGetRequest {
    repeated string keys = 1;
    string space = 2;
}

message MGetReply {
    repeated string values = 1;
}

message RuleUnit {
	string space = 1;
	uint32 index = 2;
}

message Query {
	repeated string keys = 1;
}

message SearchRequest {
	repeated RuleUnit rule = 1;
	repeated Query queries = 2;
}

message Result {
	string space = 1;
	string value = 2;
}

message SearchReply {
	repeated Result results = 1;
}
